protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,342 @@
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(DeniedUpgrade, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Some(event) = self.queued_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        if let Poll::Ready(m) = self.outbound.poll_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                ToBehaviour::TestCompleted(m.map_err(Error::Timeout).and_then(identity)),
+            ));
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
+        match event {
+            FromBehaviour::PerformRequest(req) => {
+                self.perform_request(req);
+            }
+        }
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { error, .. }) => {
+                tracing::debug!("Dial request failed: {}", error);
+                match self.queued_streams.pop_front() {
+                    Some(stream_tx) => {
+                        if let Err(_) = stream_tx.send(Err(error)) {
+                            tracing::warn!("Failed to send stream to dead handler");
+                        }
+                    }
+                    None => {
+                        tracing::warn!(
+                            "Opened unexpected substream without a pending dial request"
+                        );
+                    }
+                }
+            }
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol, ..
+            }) => match self.queued_streams.pop_front() {
+                Some(stream_tx) => {
+                    if let Err(_) = stream_tx.send(Ok(protocol)) {
+                        tracing::warn!("Failed to send stream to dead handler");
+                    }
+                }
+                None => {
+                    tracing::warn!("Opened unexpected substream without a pending dial request");
+                }
+            },
+            ConnectionEvent::RemoteProtocolsChange(ProtocolsChange::Added(mut added)) => {
+                if added.any(|p| p.as_ref() == REQUEST_PROTOCOL_NAME) {
+                    self.queued_events.push_back(
+                        ConnectionHandlerEvent::NotifyBehaviour(ToBehaviour::PeerHasServerSupport)
+                    );
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+async fn start_substream_handle(
```

Comment:
Nit: We started using the `stream` terminology as that is what libp2p as a project uses. "substream" is/was a rust-libp2p specific term that we are phasing out.

Reply:
I renamed it.